mindmeld/active_learning/heuristics.py

```diff
@@ -0,0 +1,521 @@
+from typing import List, Dict
+from collections import Counter
+from scipy.stats import entropy as scipy_entropy
+import numpy as np
+
+from ..constants import ENTROPY_LOG_BASE, ACTIVE_LEARNING_RANDOM_SEED
+
+
+def custom_reordering(confidences, do_rank: bool = True, sampling_size: int = None):
+    """
+    Args:
+        confidences: List of confidence scores from heuristic.
+        do_rank: If True, this function will return rank-sorted confidence scores.
+        sampling_size: Size of data being sampled in a turn.
+    """
+    if not do_rank:
+        assert isinstance(sampling_size, int) and sampling_size > 0, print(
+            type(sampling_size), sampling_size
+        )
+        if sampling_size >= len(confidences):
+            return np.arange(len(confidences))
+        # samples at indices {0,...,sampling_size-1}<={sampling_size}<={sampling_size+1,...}
+        idxs = np.argpartition(confidences, sampling_size)
+    else:
+        idxs = np.argsort(confidences)  # ascending order sort
+    return idxs
+
+
+class Heuristic(ABC):
+    def __init__(self, sampling_size):
+        self.sampling_size = sampling_size
+        self.check_sampling_size()
+
+    def check_sampling_size(self):
+        """Verify the sampling size is a positive integer"""
+        assert isinstance(self.sampling_size, int)
+        assert self.sampling_size > 0, print(
+            f"Current Sampling Size: {self.sampling_size}, Required size > 0"
+        )
+
+    @abstractmethod
+    def _extractor(self, **kwargs):
+        raise NotImplementedError(
+            "Subclasses must implement this method for heuristic logic."
+        )
+
+    def sample(self, **kwargs):
+        """
+        Args:
+            sampled (list, optional): List of sampled queries
+            unsampled (list, optional): List of unsampled queries
+            do_rank (bool, optional): if True, returns a ranked list of the indices for
+                confidence-sorting of data samples
+            return_dict (bool, optional): if True, return a dict with sampled, unsampled, and
+                ranked indicies.
+        """
+        sampled, unsampled, do_rank = (
+            kwargs.get("sampled", []),
+            kwargs.get("unsampled"),
+            kwargs.get("do_rank"),
+        )
+        sampled_indices, unsampled_indices, ranked_indices = self._extractor(**kwargs)
+        if kwargs.get("return_dict"):
+            return {
+                "sampled_indices": sampled_indices,
+                "unsampled_indices": unsampled_indices,
+                "ranked_indices": ranked_indices,
+            }
+        else:
+            newly_sampled = [unsampled[i] for i in sampled_indices]
+            sampled += newly_sampled
+            unsampled = [unsampled[i] for i in unsampled_indices]
+            ranked = [unsampled[i] for i in ranked_indices] if do_rank else None
+            return newly_sampled, sampled, unsampled, ranked
+
+
+class StrategicRandomSampling(Heuristic):
```

Comment:
How is this different from random sampling? A doc description here would be good

Reply:
Added the following description in the docs: Selection strategy that aims to randomly sample queries such that there is an even distribution across domains or intents.